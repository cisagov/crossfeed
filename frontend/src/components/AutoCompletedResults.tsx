import React from 'react';
import { Popover, PopoverProps, MenuItem } from '@material-ui/core';

interface Props extends PopoverProps {
  values: {
    id: { raw: string };
    text: { raw: string };
  }[];
}

export const AutoCompleteResults: React.FC<Props> = props => {
  const { values, ...rest } = props;

  return (
    <Popover {...rest}>
      {values.map(value => (
        <MenuItem key={value.id.raw}>{value.text.raw}</MenuItem>
      ))}
    </Popover>
  );
};
