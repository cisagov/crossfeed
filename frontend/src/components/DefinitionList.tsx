import React from 'react';
import { makeStyles } from '@material-ui/core';

interface Props {
  items: {
    label: string;
    value: string;
  }[];
}

export const DefinitionList: React.FC<Props> = (props) => {
  const { items } = props;
  const classes = useStyles();

  return (
    <dl className={classes.list}>
      {items.map(({ label, value }) => (
        <div className={classes.item} key={value}>
          <dt>{label}:</dt>
          <dd>{value}</dd>
        </div>
      ))}
    </dl>
  );
};

const useStyles = makeStyles((theme) => ({
  root: {
    width: '100%',
    margin: '0 0 1.5rem 0'
  },
  list: {
    width: '100%',
    margin: 0
  },
  item: {
    width: '100%',
    display: 'flex',
    flexFlow: 'row wrap',

    '& > dt': {
      color: '#71767A',
      display: 'block',
      margin: 0,
      paddingRight: '0.5rem'
    },
    '& > dd': {
      color: '#3D4551',
      margin: 0
    }
  }
}));
