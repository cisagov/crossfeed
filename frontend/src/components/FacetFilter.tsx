import React from 'react';
import {
  FormGroup,
  FormControlLabel,
  Checkbox,
  makeStyles
} from '@material-ui/core';

interface Props {
  options: { value: string; count: number }[];
  selected: string[];
  onSelect(value: string): void;
  onDeselect(value: string): void;
}

export const FacetFilter: React.FC<Props> = props => {
  const classes = useStyles();
  const { options, selected, onSelect, onDeselect } = props;

  const handleChange = (
    e: React.ChangeEvent<HTMLInputElement>,
    value: string
  ) => {
    e.persist();
    if (e.target.checked) {
      onSelect(value);
    } else {
      onDeselect(value);
    }
  };

  return (
    <>
      <FormGroup classes={{ root: classes.root }}>
        {/* <input className={classes.inp} placeholder="Filter" /> */}
        {options.map(opt => (
          <FormControlLabel
            classes={{ label: classes.label, root: classes.formControl }}
            key={opt.value}
            control={
              <Checkbox
                checked={selected.includes(opt.value)}
                onChange={e => handleChange(e, opt.value)}
              />
            }
            label={
              <>
                <span>{opt.value}</span>
                <span className={classes.count}>{opt.count}</span>
              </>
            }
          />
        ))}
      </FormGroup>
    </>
  );
};

const useStyles = makeStyles(theme => ({
  root: {
    width: '100%',
    paddingTop: 0,
    flexWrap: 'nowrap'
  },
  inp: {
    border: 'none',
    backgroundColor: '#fff',
    width: '100%',
    padding: '1rem',
    boxShadow: 'inset 0 1px 2px rgba(0,0,0,.39), 0 -1px 1px #FFF, 0 1px 0 #FFF'
  },
  label: {
    width: '100%',
    display: 'flex',
    flexFlow: 'row nowrap',
    justifyContent: 'space-between',
    alignItems: 'center',
    fontSize: '0.9rem',
    marginRight: 0,
    '& span': {
      display: 'inline-block'
    },
    '& $count': {
      fontSize: '0.7rem',
      color: theme.palette.grey[700]
    }
  },
  formControl: {
    width: 'calc(100% + 8px)'
  },
  count: {}
}));
