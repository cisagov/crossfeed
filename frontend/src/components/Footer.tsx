import React from 'react';
import { Box, Grid, Link, useTheme } from '@material-ui/core';
import { useAuthContext } from 'context';
import logo from '../assets/crossfeed.svg';

export const CrossfeedFooter: React.FC = () => {
  const { logout } = useAuthContext();
  const theme = useTheme();
  return (
    <Box style={{ backgroundColor: theme.palette.primary.main }} width={'100%'}>
      <Grid
        container
        spacing={2}
        alignItems={'center'}
        justifyContent={'center'}
        style={{
          margin: '0 auto',
          marginTop: '1rem',
          maxWidth: '1444px'
        }}
      >
        <Grid
          item
          xs={12}
          sm={2}
          style={{
            display: 'flex',
            justifyContent: 'center',
            height: '100%'
          }}
        >
          <Link href="/">
            <img src={logo} alt="Crossfeed Icon Navigate Home" />
          </Link>
        </Grid>
        <Grid
          item
          xs={12}
          sm={2}
          style={{
            display: 'flex',
            justifyContent: 'center',
            borderRight: '1px solid white',
            borderLeft: '1px solid white'
          }}
        >
          <p>
            <Link href="/" underline="always" style={{ color: 'white' }}>
              Home
            </Link>
          </p>
        </Grid>
        <Grid
          item
          xs={12}
          sm={2}
          style={{
            display: 'flex',
            justifyContent: 'center',
            borderRight: '1px solid white'
          }}
        >
          <p>
            <Link
              href="https://docs.crossfeed.cyber.dhs.gov/"
              underline="always"
              style={{ color: 'white' }}
            >
              Documentation
            </Link>
          </p>
        </Grid>
        <Grid
          item
          xs={12}
          sm={2}
          style={{
            display: 'flex',
            justifyContent: 'center',
            borderRight: '1px solid white'
          }}
        >
          <p>
            <Link
              href="https://www.cisa.gov"
              underline="always"
              style={{ color: 'white' }}
            >
              CISA Homepage
            </Link>
          </p>
        </Grid>
        <Grid
          item
          xs={12}
          sm={2}
          style={{ display: 'flex', justifyContent: 'center' }}
        >
          <p>
            <Link
              href="mailto:vulnerability@cisa.dhs.gov"
              underline="always"
              style={{ color: 'white' }}
            >
              Contact Us
            </Link>
          </p>
        </Grid>
        <Grid
          item
          xs={12}
          sm={2}
          style={{
            display: 'flex',
            justifyContent: 'center',
            borderLeft: '1px solid white'
          }}
        >
          <p>
            <Link
              href="/"
              underline="always"
              style={{ color: 'white' }}
              onClick={logout}
            >
              Logout
            </Link>
          </p>
        </Grid>
      </Grid>
    </Box>
  );
};
