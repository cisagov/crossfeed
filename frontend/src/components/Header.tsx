import React, { useState } from 'react';
import clsx from 'classnames';
import { NavLink, Link } from 'react-router-dom';
import { makeStyles } from '@material-ui/core/styles';
import {
  AppBar,
  Toolbar,
  IconButton,
  Drawer,
  ListItem,
  List
} from '@material-ui/core';
import {
  Menu as MenuIcon,
  AccountCircle as UserIcon
} from '@material-ui/icons';
import { useAuthContext } from 'context';
import logo from '../assets/cisa_logo.png';

const GLOBAL_ADMIN = 4;
const ORG_ADMIN = 2;
const ORG_USER = 1;
const ALL_USERS = GLOBAL_ADMIN | ORG_ADMIN | ORG_USER;

const useStyles = makeStyles((theme) => ({
  inner: {
    maxWidth: 1440,
    width: '100%',
    margin: '0 auto'
  },
  menuButton: {
    marginRight: theme.spacing(2),
    display: 'block',
    [theme.breakpoints.up('lg')]: {
      display: 'none'
    }
  },
  logo: {
    display: 'none',
    height: 60,
    padding: theme.spacing(),
    paddingLeft: 0,
    [theme.breakpoints.up('sm')]: {
      display: 'block'
    }
  },
  spacing: {
    flexGrow: 1
  },
  activeLink: {
    '&:after': {
      content: "''",
      position: 'absolute',
      bottom: 0,
      left: 0,
      width: '100%',
      height: 2,
      backgroundColor: 'white'
    }
  },
  activeMobileLink: {
    fontWeight: 700,
    '&:after': {
      content: "''",
      position: 'absolute',
      top: 0,
      bottom: 0,
      left: 0,
      height: '100%',
      width: 2,
      backgroundColor: theme.palette.primary.main
    }
  },
  link: {
    position: 'relative',
    color: 'white',
    textDecoration: 'none',
    margin: `0 ${theme.spacing()}px`,
    padding: theme.spacing(),
    borderBottom: '2px solid transparent',
    fontWeight: 600
  },
  userLink: {
    display: 'flex',
    alignItems: 'center',

    '& svg': {
      marginRight: theme.spacing()
    }
  },
  mdNav: {
    display: 'none',
    [theme.breakpoints.up('md')]: {
      display: 'block'
    }
  },
  lgNav: {
    display: 'none',
    [theme.breakpoints.up('lg')]: {
      display: 'block'
    }
  },
  mobileNav: {
    padding: `${theme.spacing(2)}px ${theme.spacing()}px`
  }
}));

interface NavItem {
  title: string;
  path: string;
  users: number;
}

export const Header: React.FC = () => {
  const classes = useStyles();
  const { currentOrganization, user } = useAuthContext();
  const [navOpen, setNavOpen] = useState(false);

  let userLevel = 0;
  if (user && user.isRegistered) {
    if (user.userType === 'standard') {
      if (currentOrganization?.userIsAdmin) {
        userLevel = ORG_ADMIN;
      } else {
        userLevel = ORG_USER;
      }
    } else {
      userLevel = GLOBAL_ADMIN;
    }
  }

  const navItems: NavItem[] = [
    { title: 'Inventory', path: '/', users: ALL_USERS },
    { title: 'Search', path: '/search', users: ALL_USERS },
    { title: 'Vulnerabilities', path: '/vulnerabilities', users: ALL_USERS },
    { title: 'Risk Summary', path: '/risk', users: ALL_USERS },
    {
      title: 'Organization Settings',
      path: '/organization',
      users: ORG_ADMIN | GLOBAL_ADMIN
    },
    {
      title: 'My Organizations',
      path: '/organizations',
      users: ORG_USER | ORG_ADMIN
    },
    {
      title: 'Manage Organizations',
      path: '/organizations',
      users: GLOBAL_ADMIN
    },

    { title: 'Scans', path: '/scans', users: GLOBAL_ADMIN },
    { title: 'Manage Users', path: '/users', users: GLOBAL_ADMIN }
  ].filter(({ users }) => (users & userLevel) > 0);

  const desktopNavItems: JSX.Element[] = navItems.map(({ title, path }) => (
    <NavLink
      to={path}
      key={title}
      activeClassName={classes.activeLink}
      className={classes.link}
      exact={true}
    >
      {title}
    </NavLink>
  ));

  return (
    <div>
      <AppBar position="static">
        <div className={classes.inner}>
          <Toolbar>
            <IconButton
              edge="start"
              className={classes.menuButton}
              aria-label="toggle mobile menu"
              color="inherit"
              onClick={() => setNavOpen((open) => !open)}
            >
              <MenuIcon />
            </IconButton>
            <Link to="/">
              <img
                src={logo}
                className={classes.logo}
                alt="Crossfeed Icon Navigate Home"
              />
            </Link>
            <div className={classes.mdNav}>{desktopNavItems.slice(0, 3)}</div>
            <div className={classes.lgNav}>
              {desktopNavItems.slice(3, navItems.length)}
            </div>

            <div className={classes.spacing} />
            {userLevel > 0 && (
              <NavLink
                to="/settings"
                activeClassName={classes.activeLink}
                className={clsx(classes.link, classes.userLink)}
                exact={true}
              >
                <UserIcon /> My Account
              </NavLink>
            )}
          </Toolbar>
        </div>
      </AppBar>
      <Drawer
        anchor="left"
        open={navOpen}
        onClose={() => setNavOpen(false)}
        data-testid="mobilenav"
      >
        <List className={classes.mobileNav}>
          {navItems.map(({ title, path }) => (
            <ListItem
              button
              key={title}
              component={NavLink}
              to={path}
              activeClassName={classes.activeMobileLink}
              exact={true}
            >
              {title}
            </ListItem>
          ))}
        </List>
      </Drawer>
    </div>
  );
};
