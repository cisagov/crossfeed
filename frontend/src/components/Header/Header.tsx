import React, { useState } from 'react';
import { NavLink, Link } from 'react-router-dom';
import {
  Header as UsaHeader,
  NavMenuButton,
  Title,
  PrimaryNav
} from '@trussworks/react-uswds';
import logo from './cisa_logo.png';
import classes from './styles.module.scss';
import { useAuthContext } from 'context';

export const Header: React.FC = () => {
  const { currentOrganization, user } = useAuthContext();
  const [mobileExpanded, setMobileExpanded] = useState(false);

  const onExpandMobile: React.MouseEventHandler<HTMLButtonElement> = e => {
    setMobileExpanded(expanded => !expanded);
  };

  const dashboardTitle = 'Dashboard';

  const orgUserNav = [
    <NavLink
      activeClassName="usa-current"
      to="/"
      exact
      key="dashboard"
      className="usa-nav__link"
    >
      <span>{dashboardTitle}</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/vulnerabilities"
      key="vulnerabilities"
      className="usa-nav__link"
    >
      <span>Vulnerabilities</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/risk"
      key="risk"
      className="usa-nav__link"
    >
      <span>Risk Summary</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/organizations"
      key="organizations"
      className="usa-nav__link"
    >
      <span>My Organizations</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/settings"
      key="settings"
      className="usa-nav__link"
    >
      <span>My Account</span>
    </NavLink>
  ];

  const orgAdminNav = [
    <NavLink
      activeClassName="usa-current"
      to="/"
      exact
      key="dashboard"
      className="usa-nav__link"
    >
      <span>{dashboardTitle}</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/vulnerabilities"
      key="vulnerabilities"
      className="usa-nav__link"
    >
      <span>Vulnerabilities</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/risk"
      key="risk"
      className="usa-nav__link"
    >
      <span>Risk Summary</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/organization"
      key="organization"
      className="usa-nav__link"
    >
      <span>Organization Settings</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/organizations"
      key="organizations"
      className="usa-nav__link"
    >
      <span>My Organizations</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/scans"
      key="scans"
      className="usa-nav__link"
    >
      <span>Scans</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/settings"
      key="settings"
      className="usa-nav__link"
    >
      <span>My Account</span>
    </NavLink>
  ];

  const globalAdminNav = [
    <NavLink
      activeClassName="usa-current"
      to="/"
      exact
      key="dashboard"
      className="usa-nav__link"
    >
      <span>{dashboardTitle}</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/vulnerabilities"
      key="vulnerabilities"
      className="usa-nav__link"
    >
      <span>Vulnerabilities</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/risk"
      key="risk"
      className="usa-nav__link"
    >
      <span>Risk Summary</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/organization"
      key="organization"
      className="usa-nav__link"
    >
      <span>Organization Settings</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/scans"
      key="scans"
      className="usa-nav__link"
    >
      <span>All Scans</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/organizations"
      key="organizations"
      className="usa-nav__link"
    >
      <span>Manage Organizations</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/users"
      key="users"
      className="usa-nav__link"
    >
      <span>Manage Users</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/settings"
      key="settings"
      className="usa-nav__link"
    >
      <span>My Account</span>
    </NavLink>
  ];

  let nav: JSX.Element[] = [];
  if (user && user.isRegistered) {
    if (user.userType === 'standard') {
      if (currentOrganization?.userIsAdmin) {
        nav = orgAdminNav;
      } else {
        nav = orgUserNav;
      }
    } else {
      nav = globalAdminNav;
    }
  }

  return (
    <UsaHeader basic className={classes.root}>
      <div className="usa-nav-container">
        <div className="usa-navbar">
          <Link to="/">
            <Title>
              <img src={logo} alt="Crossfeed logo" />
            </Title>
          </Link>
          <NavMenuButton label="Menu" onClick={onExpandMobile} />
        </div>

        <PrimaryNav
          mobileExpanded={mobileExpanded}
          onToggleMobileNav={onExpandMobile}
          items={nav}
        />
      </div>
    </UsaHeader>
  );
};
