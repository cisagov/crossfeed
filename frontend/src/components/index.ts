export * from './Layout';
export * from './GovBanner';
export * from './Header';
export * from './AuthRoute';
export * from './Table';
export * from './Paginator';
export * from './ColumnFilter';
export * from './SelectFilter';
export * from './AuthForm';
export * from './FileInput';
export * from './ImportExport';
export * from './RouteGuard';
export * from './SearchBar';
export * from './MTable';
export * from './TablePaginationActions';
export * from './TaggedInput';
export * from './DomainDetails';
export * from './AutoCompletedResults';
export * from './FacetFilter';
export * from './Subnav';
