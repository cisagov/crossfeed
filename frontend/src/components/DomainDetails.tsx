import React, { useCallback, useEffect, useMemo, useState } from 'react';
import { Link } from 'react-router-dom';
import {
  Paper,
  makeStyles,
  Accordion,
  AccordionSummary,
  Typography,
  AccordionDetails,
  List,
  ListItem,
  ListItemText,
  Collapse
} from '@material-ui/core';
import {
  ExpandLess,
  ExpandMore,
  Launch as LinkOffIcon
} from '@material-ui/icons';
import { Domain } from 'types';
import { useDomainApi } from 'hooks';
import { DefinitionList } from './DefinitionList';
import { formatDistanceToNow, parseISO } from 'date-fns';
import { stateMap } from 'pages/Vulnerabilities/Vulnerabilities';
import { Webpage } from 'types/webpage';
import { useAuthContext } from 'context';

interface Props {
  domainId: string;
}

export const generateWebpageTree = (pages: Webpage[]) => {
  const tree: any = {};
  for (const page of pages) {
    const url = new URL(page.url);
    const parts = url.pathname.split('/').filter((path) => path !== '');
    let root = tree;
    for (let i = 0; i < parts.length - 1; i++) {
      if (parts[i] in root) root = root[parts[i]];
      else {
        root[parts[i]] = {};
        root = root[parts[i]];
      }
    }
    root[parts[parts.length - 1]] = page;
  }
  return tree;
};

export const DomainDetails: React.FC<Props> = (props) => {
  const { domainId } = props;
  const { getDomain } = useDomainApi(false);
  const { user } = useAuthContext();
  const [domain, setDomain] = useState<Domain>();
  const classes = useStyles();

  const fetchDomain = useCallback(async () => {
    try {
      setDomain(undefined);
      const result = await getDomain(domainId);
      setDomain(result);
    } catch (e) {
      console.error(e);
    }
  }, [domainId, getDomain]);

  useEffect(() => {
    fetchDomain();
  }, [fetchDomain]);

  const webInfo = useMemo(() => {
    if (!domain) {
      return [];
    }
    const categoriesToProducts: Record<string, Set<string>> = {};
    for (const service of domain.services) {
      for (const product of service.products) {
        const version = product.version ? ` ${product.version}` : '';
        const value = product.name + version;
        const name =
          product.tags && product.tags.length > 0 ? product.tags[0] : 'Misc';
        if (!categoriesToProducts[name]) {
          categoriesToProducts[name] = new Set();
        }
        categoriesToProducts[name].add(value);
      }
    }
    return Object.entries(categoriesToProducts).reduce(
      (acc, [name, value]) => [
        ...acc,
        {
          label: name,
          value: Array.from(value).join(', ')
        }
      ],
      [] as any
    );
  }, [domain]);

  const overviewInfo = useMemo(() => {
    if (!domain) {
      return [];
    }
    const ret = [];
    if (domain.ip) {
      ret.push({
        label: 'IP',
        value: domain.ip
      });
    }
    ret.push({
      label: 'First Seen',
      value: `${formatDistanceToNow(parseISO(domain.createdAt))} ago`
    });
    ret.push({
      label: 'Last Seen',
      value: `${formatDistanceToNow(parseISO(domain.updatedAt))} ago`
    });
    if (domain.country) {
      ret.push({
        label: 'Country',
        value: domain.country
      });
    }
    if (domain.cloudHosted) {
      ret.push({
        label: 'Cloud Hosted',
        value: 'Yes'
      });
    }
    ret.push({
      label: 'Organization',
      value: domain.organization.name
    });
    return ret;
  }, [domain]);

  const [hiddenRows, setHiddenRows] = React.useState<{
    [key: string]: boolean;
  }>({});

  const formatBytes = (bytes: number, decimals = 2): string => {
    if (bytes === 0) return '0 Bytes';

    const k = 1024;
    const dm = decimals < 0 ? 0 : decimals;
    const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];

    const i = Math.floor(Math.log(bytes) / Math.log(k));

    return parseFloat((bytes / Math.pow(k, i)).toFixed(dm)) + ' ' + sizes[i];
  };

  const generateWebpageList = (tree: any, prefix = '') => {
    return (
      <List
        className={`${classes.listRoot}${prefix ? ' ' + classes.nested : ''}`}
      >
        {Object.keys(tree).map((key) => {
          const isWebpage =
            'url' in tree[key] && typeof tree[key]['url'] === 'string';
          if (!isWebpage) {
            let newPrefix = prefix + '/' + key;
            return (
              <>
                <ListItem
                  button
                  onClick={() => {
                    setHiddenRows((hiddenRows: any) => {
                      hiddenRows[newPrefix] =
                        newPrefix in hiddenRows ? !hiddenRows[newPrefix] : true;
                      return { ...hiddenRows };
                    });
                  }}
                  key={newPrefix}
                >
                  <ListItemText primary={(prefix ? '' : '/') + key + '/'} />
                  {hiddenRows[newPrefix] ? <ExpandLess /> : <ExpandMore />}
                </ListItem>
                <Collapse
                  in={!hiddenRows[newPrefix]}
                  timeout="auto"
                  unmountOnExit
                >
                  {generateWebpageList(tree[key], newPrefix)}
                </Collapse>
              </>
            );
          }
          const page = tree[key] as Webpage;
          const parsed = new URL(page.url);
          const split = parsed.pathname
            .replace(/\/$/, "") // Remove trailing slash
            .split('/');
          return (
            <ListItem
              button
              divider={true}
              key={page.url}
              onClick={() => window.open(page.url, '_blank')}
            >
              <ListItemText
                primary={(prefix ? '' : '/') + split.pop()}
                secondary={
                  page.status + ' â€¢ ' + formatBytes(page.responseSize ?? 0, 1)
                }
              ></ListItemText>
            </ListItem>
          );
        })}
      </List>
    );
  };

  if (!domain) {
    return null;
  }

  const url =
    (domain.services.find((service) => service.port === 443)
      ? 'https://'
      : 'http://') + domain.name;

  domain.webpages.sort((a, b) => (a.url > b.url ? 1 : -1));
  const webpageTree = generateWebpageTree(domain.webpages);
  const webpageList = generateWebpageList(webpageTree);

  return (
    <Paper classes={{ root: classes.root }}>
      <div className={classes.title}>
        <h4>
          <Link to={`/domain/${domain.id}`}>{domain.name}</Link>
        </h4>

        <a href={url} target="_blank" rel="noopener noreferrer">
          <LinkOffIcon />
        </a>
      </div>
      <div className={classes.inner}>
        {overviewInfo.length > 0 && (
          <div className={classes.section}>
            <h4 className={classes.subtitle}>Overview</h4>
            <DefinitionList items={overviewInfo} />
          </div>
        )}
        {webInfo.length > 0 && (
          <div className={classes.section}>
            <h4 className={classes.subtitle}>Known Products</h4>
            <DefinitionList items={webInfo} />
          </div>
        )}

        {domain.vulnerabilities.length > 0 && (
          <div className={classes.section}>
            <h4 className={classes.subtitle}>Vulnerabilities</h4>
            <Accordion className={classes.accordionHeaderRow} disabled>
              <AccordionSummary expandIcon={<ExpandMore />}>
                <Typography className={classes.accordionHeading}>
                  Title
                </Typography>
                <Typography className={classes.vulnDescription}>
                  Serverity
                </Typography>
                <Typography className={classes.vulnDescription}>
                  State
                </Typography>
                <Typography className={classes.vulnDescription}>
                  Created
                </Typography>
              </AccordionSummary>
            </Accordion>
            {domain.vulnerabilities.map((vuln) => (
              <Accordion className={classes.accordion} key={vuln.id}>
                <AccordionSummary expandIcon={<ExpandMore />}>
                  <Typography className={classes.accordionHeading}>
                    {vuln.cve}
                  </Typography>
                  <Typography className={classes.vulnDescription}>
                    {vuln.severity}
                  </Typography>
                  <Typography className={classes.vulnDescription}>
                    {vuln.state}
                  </Typography>
                  <Typography className={classes.vulnDescription}>
                    {vuln.createdAt
                      ? `${formatDistanceToNow(parseISO(vuln.createdAt))} ago`
                      : ''}
                  </Typography>
                </AccordionSummary>
                <AccordionDetails>
                  <DefinitionList
                    items={[
                      {
                        label: 'CVE',
                        value: vuln.cve ?? 'N/A'
                      },
                      {
                        label: 'Severity',
                        value: vuln.severity ?? 'N/A'
                      },
                      {
                        label: 'CVSS',
                        value: vuln.cvss?.toString() ?? 'N/A'
                      },
                      {
                        label: 'CPE',
                        value: vuln.cpe ?? 'N/A'
                      },
                      {
                        label: 'State',
                        value:
                          `${vuln.state} (${stateMap[
                            vuln.substate
                          ].toLowerCase()})` ?? 'N/A'
                      },
                      {
                        label: 'Description',
                        value: vuln.description ?? 'N/A'
                      }
                    ]}
                  />
                </AccordionDetails>
              </Accordion>
            ))}
          </div>
        )}
        {domain.services.length > 0 && (
          <div className={classes.section}>
            <h4 className={classes.subtitle}>Ports</h4>
            <Accordion className={classes.accordionHeaderRow} disabled>
              <AccordionSummary expandIcon={<ExpandMore />}>
                <Typography className={classes.accordionHeading}>
                  Port
                </Typography>
                <Typography className={classes.accordionHeading}>
                  Products
                </Typography>
                <Typography>Last Seen</Typography>
              </AccordionSummary>
            </Accordion>
            {domain.services.map((service) => {
              const products = service.products
                .map(
                  (product) =>
                    product.name +
                    (product.version ? ` ${product.version}` : '')
                )
                .join(', ');
              return (
                <Accordion className={classes.accordion} key={service.id}>
                  <AccordionSummary expandIcon={<ExpandMore />}>
                    <Typography className={classes.accordionHeading}>
                      {service.port}
                    </Typography>
                    <Typography className={classes.accordionHeading}>
                      {products}
                    </Typography>
                    {service.lastSeen && (
                      <Typography>
                        {formatDistanceToNow(parseISO(service.lastSeen))} ago
                      </Typography>
                    )}
                  </AccordionSummary>
                  {service.products.length > 0 && (
                    <AccordionDetails>
                      <DefinitionList
                        items={[
                          {
                            label: 'Products',
                            value: products
                          },
                          {
                            label: 'Banner',
                            value:
                              (user?.userType === 'globalView' ||
                                user?.userType === 'globalAdmin') &&
                              service.banner
                                ? service.banner
                                : 'None'
                          }
                        ]}
                      />
                    </AccordionDetails>
                  )}
                </Accordion>
              );
            })}
          </div>
        )}
        {domain.webpages.length > 0 && (
          <div className={classes.section}>
            <h4 className={classes.subtitle}>Site Map</h4>
            {webpageList}
          </div>
        )}
      </div>
    </Paper>
  );
};

const useStyles = makeStyles((theme) => ({
  root: {
    border: `2px solid ${theme.palette.primary.main}`,
    boxShadow: '0px 1px 6px rgba(0, 0, 0, 0.25)',
    marginBottom: '1rem',
    '& *:focus': {
      outline: 'none !important'
    }
  },
  title: {
    backgroundColor: theme.palette.primary.main,
    color: '#fff',
    cursor: 'pointer',
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    padding: '1rem 1.5rem',
    fontSize: '2rem',
    textDecoration: 'none',

    '& > h4': {
      wordBreak: 'break-all',
      paddingRight: '2rem',
      margin: '0'
    },

    '& > a, & > h4 a': {
      color: 'white',
      textDecoration: 'none'
    }
  },
  section: {
    marginBottom: '1.5rem'
  },
  subtitle: {
    margin: 0,
    padding: '0 0 0.2rem 0',
    fontSize: '1.2rem',
    fontWeight: 500,
    color: '#3D4551'
  },
  inner: {
    padding: '1.5rem'
  },
  accordion: {
    color: '#3D4551'
  },
  accordionHeaderRow: {
    color: '#000',
    backgroundColor: '#eaeaea !important'
  },
  accordionHeading: {
    flex: '1 0 33%'
  },
  vulnDescription: {
    flex: '1 1 15%',
    textOverflow: 'hidden',
    textAlign: 'right'
  },
  listRoot: {
    width: '100%',
    backgroundColor: theme.palette.background.paper
  },
  nested: {
    paddingLeft: theme.spacing(2)
  }
}));
