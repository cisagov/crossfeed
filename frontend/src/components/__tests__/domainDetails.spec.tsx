import { generateWebpageTree } from 'components/DomainDetails';
import React from 'react';
import { render, fireEvent, testUser, testOrganization } from 'test-utils';
import { Header } from '../Header';

describe('generateWebpageTree', () => {
    it('no path', () => {
        const pages = [
            {
                url: "https://www.cisa.gov/"
            }
        ];
        const tree = generateWebpageTree(pages as any);
        expect(tree).toMatchSnapshot();
    });
    it('basic path', () => {
        const pages = [
            {
                url: "https://www.cisa.gov/a/b/c"
            }
        ];
        const tree = generateWebpageTree(pages as any);
        expect(tree).toMatchSnapshot();
    });
    it('single path', () => {
        const pages = [
            {
                url: "https://www.cisa.gov/a"
            }
        ];
        const tree = generateWebpageTree(pages as any);
        expect(tree).toMatchSnapshot();
    });
    it('trailing slash', () => {
        const pages = [
            {
                url: "https://www.cisa.gov/a/"
            }
        ];
        const tree = generateWebpageTree(pages as any);
        expect(tree).toMatchSnapshot();
    });
});