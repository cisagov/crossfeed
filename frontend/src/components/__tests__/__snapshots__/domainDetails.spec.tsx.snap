// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateWebpageTree basic path 1`] = `
Object {
  "a": Object {
    "b": Object {
      "c": Object {
        "url": "https://www.cisa.gov/a/b/c",
      },
    },
  },
}
`;

exports[`generateWebpageTree no path 1`] = `
Object {
  "undefined": Object {
    "url": "https://www.cisa.gov/",
  },
}
`;

exports[`generateWebpageTree single path 1`] = `
Object {
  "a": Object {
    "url": "https://www.cisa.gov/a",
  },
}
`;

exports[`generateWebpageTree trailing slash 1`] = `
Object {
  "a": Object {
    "url": "https://www.cisa.gov/a/",
  },
}
`;
